src/utils/utils_log.c

```diff
@@ -0,0 +1,204 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#define _GNU_SOURCE 1
+#include <stdarg.h>
+#include <string.h>
+#include <time.h>
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <sys/types.h>
+#include <unistd.h>
+#endif
+
+#include "utils_common.h"
+#include "utils_log.h"
+
+typedef struct {
+    int timestamp;
+    int pid;
+    LogLevel level;
+    LogLevel flushLevel;
+    FILE *output;
+} LoggerConfig;
+
+LoggerConfig loggerConfig = {0, 0, LOG_ERROR, LOG_ERROR, NULL};
+
+static const char *levelToStr(LogLevel l) {
+    switch (l) {
+    case LOG_DEBUG:
+        return "DEBUG";
+    case LOG_ERROR:
+        return "ERROR";
+    case LOG_INFO:
+        return "INFO";
+    case LOG_WARNING:
+        return "WARN";
+    default:
+        ASSERT(0);
+    }
+}
+static const char *matchEnvVar(const char *envVar, const char *prefix,
+                               const char **extraArg) {
+    const char *found = strstr(envVar, prefix);
+    // ensure that found string is a first on list or it's separated semicolon
+    if (!found) {
+        return NULL;
+    }
+    // if found string is not at the begining of envVar ensure if it's preceded by ';'
+    if (found != envVar && *(found - 1) != ';') {
+        return NULL;
+    }
+
+    const char *endFound = found + strlen(prefix);
+    if (!extraArg) {
+        // if there is no argument matched string should end with ';' or '\0'
+        if (*endFound != '\0' && *endFound != ';') {
+            return NULL;
+        }
+        return found;
+    }
+
+    // check if matched string ends with '='
+    if (*endFound != '=') {
+        return NULL;
+    }
+
+    *extraArg = endFound + 1;
+
+    return found;
+}
+
+void utils_log(LogLevel level, const char *format, ...) {
+
+    if (!loggerConfig.output) {
+        loggerConfig.output = stderr;
+    }
+    if (level < loggerConfig.level) {
+        return;
+    }
+
+#ifdef _WIN32
+    DWORD pid = GetCurrentProcessId();
+    DWORD tid = GetCurrentThreadId();
+#else
+    pid_t pid = getpid();
+    pid_t tid = gettid();
+#endif
+
+    char buffer[2048];
+    va_list args;
+    va_start(args, format);
+    int ret = vsnprintf(buffer, sizeof(buffer), format, args);
+    const char *overflow = "";
+    if (ret >= (intptr_t)sizeof(buffer)) {
+        //TODO: alloc bigger buffer with base alloc
+        overflow = "[truncated...]";
+    }
+    va_end(args);
+    char header[256];
```

Comment:
please add a #define for this and all magic numbers and move these #defines to the top of the file

Reply:
i added those defines. This is matter of codding style, but i proffer keep defines which are used only in the single function next to it definition. It's improves code readability, when you can see define and it's usage on single screen.